### ISSUE TITLE
keywords if/then/else require v5 option

### ISSUE THREAD

**Issue by egoist** [webpack/webpack#3562]
the ajv-keywords@^1.3.0 module added new keyword https://github.com/epoberezkin/ajv-keywords/blob/master/keywords/if.js#L4

**Do you want to request a *feature* or report a *bug*?**
maybe a bug

**What is the current behavior?**
console would log `keywords if/then/else require v5 option`

**What is the expected behavior?**
no such log

**Please mention other relevant information such as the browser version, Node.js version, Operating System and programming language.**
webpack@^2.2.0-rc.1

**Comment by sokra** [comment]
Sounds reasonable. Do you want to provide a PR?

**Comment by just-boris** [comment]
I believe that it will be fixed by @epoberezkin in the https://github.com/epoberezkin/ajv-keywords/issues/5

**Comment by dschissler** [comment]
I'm getting this message on `2.2.0-rc.2`.  Can someone explain what this means?

**Comment by noherczeg** [comment]
As mentioned before the problem still exists in `2.2.0-rc.2`. Is this an actual error?

**Comment by TheLarkInn** [comment]
Have you reviewed the merged pull request which I closed this issue for? 😏

**Comment by d-ph** [comment]
tl;dr: Error will go away in the next release after `2.2.0-rc.2`. The fix: https://github.com/webpack/webpack/pull/3566/files

**Comment by PTC-JoshuaMatthews** [comment]
Can someone help me understand how this issue can just start happening out of nowhere? I had everything working last week on my build server, no code has been changed since, and all of the sudden my builds are failing with this error. I don't understand, since I have specific versions of libraries in my package.json how can issues like this be affecting me without making any changes?

**Comment by d-ph** [comment]
This warning is not specific to webpack. It can be displayed by any code/library, that uses `ajv-keywords` lib. It's likely, that you started to see this warning, because of some other js library (mis)using `ajv-keywords`. 

And unless you use [npm shrinkwrap](https://docs.npmjs.com/cli/shrinkwrap), you might get newer versions of deeper libraries even though you might have your immediate libraries locked in your `package.json`.

Just out of curiosity, could I see what gets output, when you run `npm ls webpack`, please?

**Comment by epoberezkin** [comment]
@d-ph the issue is closed and fixed in webpack - it uses ajv-keywords directly.

**Comment by PTC-JoshuaMatthews** [comment]
@d-ph ` npm ls webpack `displays `webpack@2.1.0-beta.27`

**Comment by noherczeg** [comment]
@PTC-JoshuaMatthews Could be caused by a transitive dependency as well.

**Comment by d-ph** [comment]
@PTC-JoshuaMatthews Ok. 

The short story is: start using npm shrinkwrap.

The long story is:
If you don't use `npm shrinkwrap`, then your `node_modules` is not reproducible. I.e. every time you build it's potentially different. That's why even though you don't make any changes in your code, things are still different every time you `npm install`. As to why `shrinkwrap` isn't the default npm's behaviour -- don't ask me.

`ajv-keywords` started displaying this warning [6 days ago](https://github.com/epoberezkin/ajv-keywords/commit/a4edf60806d37cc7772a8704f8dc08dd7977441e). So since that moment your project started displaying that message and, as you said, breaking your build.

There are different solutions for you atm. In the nutshell: get rid of the warning in any way (e.g. lock `ajv-keywords` version to one before [this commit](https://github.com/epoberezkin/ajv-keywords/commit/a4edf60806d37cc7772a8704f8dc08dd7977441e)) and `npm shrinkwrap`. After that, your project is always restorable to a known state via `npm install`.

**Comment by eltiare** [comment]
Or maybe `yarn` instead? Supposedly it's more reliable than `npm shrinkwrap`.

**Comment by teonik** [comment]
Message went away with 2.2.0-rc.3....Thanks

**Comment by syedfarrukhqamar** [comment]
Dear All, I have tried with above suggestion of npm shrinkwrap but still getting the following error

keywords if/then/else require v5 option
module.js:472
    throw err;
    ^

Error: Cannot find module 'reflect-metadata'
    at Function.Module._resolveFilename (module.js:470:15)
    at Function.Module._load (module.js:418:25)
    at Module.require (module.js:498:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/farrukhqamar/HybridApps/testing/SLife/node_modules/@ionic/app-scripts/dist/aot/aot-compiler.js:4:1)
    at Module._compile (module.js:571:32)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)

any suggestion

**Comment by d-ph** [comment]
`keywords if/then/else require v5 option` is not your error. `Error: Cannot find module 'reflect-metadata'` is. Solve it by `npm install --save reflect-metadata`

**Comment by syedfarrukhqamar** [comment]
Thanks for the answer.
now it is giving the error below
when i tried to build project for ios. 

bmc:apiapp myuser$ ionic build ios

> ionic-hello-world@ ionic:build /Users/________/HybridApps/testing/apiApp
> ionic-app-scripts build

keywords if/then/else require v5 option
[13:01:28]  ionic-app-scripts 0.0.47 
[13:01:28]  build dev started ... 
[13:01:28]  clean started ... 
[13:01:28]  clean finished in 7 ms 
[13:01:28]  copy started ... 
[13:01:28]  transpile started ... 
[13:01:31]  build dev failed: Cannot read property 'indexOf' of undefined 
[13:01:31]  ionic-app-script task: "build" 
[13:01:31]  TypeError: Cannot read property 'indexOf' of undefined 

npm ERR! Darwin 16.3.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "ionic:build" "--"
npm ERR! node v7.3.0
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! ionic-hello-world@ ionic:build: `ionic-app-scripts build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the ionic-hello-world@ ionic:build script 'ionic-app-scripts build'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the ionic-hello-world package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     ionic-app-scripts build
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs ionic-hello-world
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls ionic-hello-world
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/_________/HybridApps/testing/apiApp/npm-debug.log

**Comment by syedfarrukhqamar** [comment]
thanks a lot for all your help. now it is fixed. i have updated my ionic now and last error seems to be fixed 👍
